### PR TITLE
Added Relation/Binary/HeterogeneousEquality/Quotients.agda

### PR DESCRIPTION
**Pull request by jmchapman**
It contains a definition of quotients as a record that you can postulate
as you would for extensionality and some useful lemmas. Also added a
hetergeneous version of proof irrelevance to
HeterogeneousEquality.agda.

### CODE REVIEW

**Comment by gallais** [comment]
Looks like the build for [the refactored version of the PR](https://github.com/agda/agda-stdlib/compare/master...a65e0cfa551e) was successful: https://travis-ci.org/agda/agda-stdlib/builds/318377364.

Any volunteer for a review?

File: src/Relation/Binary/HeterogeneousEquality.agda [hunk 1]
```diff
@@ -233,3 +233,15 @@ data Reveal_is_ {a} {A : Set a} (x : Hidden A) (y : A) : Set a where
 inspect : ∀ {a b} {A : Set a} {B : A → Set b}
           (f : (x : A) → B x) (x : A) → Reveal (hide f x) is (f x)
 inspect f x = [ refl ]
+
+------------------------------------------------------------------------
+-- Heterogeneous proof irrelevance
+
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/HeterogeneousEquality.agda]: These two properties could have more meaningful names? Maybe something like `≅-irrelevance` to match `≡-irrelevance` in 'PropositionalEquality'.

File: src/Relation/Binary/HeterogeneousEquality/Quotients.agda [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    ≅⟨ liftCong p x ⟩
+    lift (g ∘ abs) (cong g ∘ absCompat) x
+    ≅⟨ isLift g x ⟩
+    g x
+    ∎
+
+--
+postulate ext : ∀{a b}{A : Set a}{B B' : A → Set b}{f : ∀ a → B a}
+                {g : ∀ a → B' a} → (∀ a → f a ≅ g a) → f ≅ g
+
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/HeterogeneousEquality/Quotients.agda]: Maybe we should consider moving everything that requires extensionality into its own module `Quotient.Extensionality`? Hopefully then the postulate would be less likely to "contaminate" other parts of the library in any future '--safe' checks...

File: src/Relation/Binary/HeterogeneousEquality.agda (hunk 1 above)
Comment (gallais) [reply src/Relation/Binary/HeterogeneousEquality.agda]: I've cleaned this up in the merge commit: https://github.com/agda/agda-stdlib/commit/e5ae09d93c8a87bc591bc468fb911b9e64ae3aa5#diff-a07ef1a4ca3de2d7c7136b83a5c15b34R96

Comment (gallais) [reply src/Relation/Binary/HeterogeneousEquality.agda]: Btw, you can see the cumulated effects of all the commits on that branch here: https://github.com/agda/agda-stdlib/compare/quotients

Comment (MatthewDaggitt) [reply src/Relation/Binary/HeterogeneousEquality.agda]: Ah, sorry hadn't seen that. Thanks for the link!

**Comment by MatthewDaggitt** [comment]
@jmchapman thank you very much for your PR. We've incorporated your suggested modules along with @gallais's changes in commit 7400d8c.

Apologies that it took quite so long to get round to it!

**Pull request closed without merging**